lib/queryBuilder/graphUpserter/UpsertGraph.js

```diff
@@ -66,64 +66,74 @@ class UpsertGraph {
   }
 
   buildGraph(current) {
-    this.doBuildGraph(this.rootModelClass, this.upsert, current, null, this.relExpr);
+    this.doBuildGraph(this.rootModelClass, this.upsert, current, null, this.relExpr, []);
   }
 
-  doBuildGraph(modelClass, upsert, current, parentNode, relExpr) {
+  doBuildGraph(modelClass, upsert, current, parentNode, relExpr, dataPath) {
     this.buildGraphArray(
       modelClass,
       ensureArray(upsert),
       ensureArray(current),
       parentNode,
-      relExpr
+      relExpr,
+      dataPath
     );
   }
 
-  buildGraphArray(modelClass, upsert, current, parentNode, relExpr) {
+  buildGraphArray(modelClass, upsert, current, parentNode, relExpr, dataPath) {
     const idProp = modelClass.getIdPropertyArray();
 
     const upsertById = keyBy(upsert, model => model.$propKey(idProp));
     const currentById = keyBy(current, model => model.$propKey(idProp));
 
-    upsert.forEach(upsert => {
+    upsert.forEach((upsert, index) => {
       const key = upsert.$propKey(idProp);
       const current = currentById[key];
 
-      this.buildGraphSingle(modelClass, upsert, current, parentNode, relExpr);
+      this.buildGraphSingle(modelClass, upsert, current, parentNode, relExpr, [...dataPath, index]);
     });
 
     current.forEach(current => {
       const key = current.$propKey(idProp);
       const upsert = upsertById[key];
 
       if (!upsert) {
-        this.buildGraphSingle(modelClass, upsert, current, parentNode, relExpr);
+        // TODO: What to do about dataPath here?
+        this.buildGraphSingle(modelClass, upsert, current, parentNode, relExpr, null);
```

Comment:
@koskimas I am now building the `dataPath` recursively, in the same way it is done for inserts. the problem here is that for a `current` model that has no corresponding `upsert` model, there isn't a correct index for it into the graph, since it doesn't actually show up in the graph. This shouldn't cause any validation errors, but what if it does? What kind of index should we give it? `-1`?

Reply:
These nodes create delete and unrelate operations and nothing is validated. I think -1 is fine here. It should never actually get used.

Reply:
Cool, I'll adjust that then.